File: b_rank_new_level_up_problems/b_rank_new_level_up_problems__grid_numbering.cs
// 実行時間 20ms
using static System.Console;
using static System.Linq.Enumerable;

internal class Program
{
    static void Main()
    {
        int c = 0;
        int x = 0;
        int y = 0;
        int[] hwd = ReadLine().Split( ' ' ).Select( int.Parse ).ToArray();
        int[,] m = new int[hwd[0], hwd[1]];

        while( c < hwd[0] * hwd[1] ) {
            if( 0 <= y && y < hwd[0] && 0 <= x && x < hwd[1] ) {
                m[y, x] = ++c;
            }
            switch( hwd[2] ) {
            case 1:
                y--;
                x++;
                if( y < 0 ) {
                    y = y + x + 1;
                    x = 0;
                }
                break;
            case 2:
                x++;
                if( x == hwd[1] ) {
                    y++;
                    x = 0;
                }
                break;
            case 3:
                y++;
                if( y == hwd[0] ) {
                    y = 0;
                    x++;
                }
                break;
            case 4:
                y++;
                x--;
                if( x < 0 ) {
                    x = y + x + 1;
                    y = 0;
                }
                break;
            }
        }
        foreach( int i in Range( 0, hwd[0] ) ) {
            int[] t = new int[hwd[1]];

            foreach( int j in Range( 0, hwd[1] ) ) {
                t[j] = m[i, j];
            }
            WriteLine( string.Join( " ", t ) );
        }
        return;
    }
}

File: b_rank_new_level_up_problems/b_rank_new_level_up_problems__grid_numbering.py
# 実行時間 20ms
h, w, d = list(map(int, input().split()))
a = [[0 for i in range(w)] for j in range(h)]
n = 0
y = 0
x = 0
l = []

while n < h * w:
    if 0 <= y < h and 0 <= x < w:
        n += 1
        a[y][x] = n
    if d == 1:
        y -= 1
        x += 1
        if y < 0:
            y = y + x + 1
            x = 0
    elif d == 2:
        x += 1
        if x == w:
            y += 1
            x = 0
    elif d == 3:
        y += 1
        if y == h:
            y = 0
            x += 1
    elif d == 4:
        y += 1
        x -= 1
        if x < 0:
            x = y + x + 1
            y = 0
for i in range(h):
    t = []

    for j in range(w):
        t.append(str(a[i][j]))
    l.append(" ".join(t))
print("\n".join(l))
